service/dap/server.go

```diff
@@ -819,16 +819,17 @@ func (s *Server) onLaunchRequest(request *dap.LaunchRequest) {
 		}
 
 		buildFlags := args.BuildFlags
+		buildDir := args.BuildDir
```

Comment:
Why not just `if args.BuildDir != "" { os.Chdir(args.BuildDir) }`?

Reply:
That limits DAP package to work only with one debug session and potentially adds further confusion to handling of the relative paths.

Reply:
> That limits DAP package to work only with one debug session

I was looking at this for #2551, there are very deeply rooted limitations around this. It might look like you can just instantiate multiple instances of Debugger but they won't work except in select circumstances.

> potentially adds further confusion to handling of the relative paths.

I was actually thinking the opposite. The PR, as it is, adds one more possible directory that a relative path can be relative to.

Reply:
Good to know that such limitation already exists. Adding one more blocker doesn't change the game.

My original plan was to interpret all the paths as paths relative to the workspace root because that's what we tell users. 
os.Chdir makes relative paths in `cwd`. But I guess vscode-go can avoid this by always explicitly setting the cwd.

I left the output to be computed based on the original working directory path for now. Should it be relative to the updated process working directory too? It's a temporary file, so either seems fine to me.

Reply:
Because of test failures switched to recover to the original wd after build. 
(debug server is running in process, so we need to make sure to switch back to the original directory for each test. I figured it's easier to just affect only build step).
But still contained changes to only service/dap package.

Once build is done, cwd is going to be used. So, keeping the original working directory shouldn't be confusing.

Reply:
If you do it this way you might as well just pass it inside exec.Command. The point of doing the `os.Chdir` would be to change to it as early as possible (so it would be the first thing you do after getting the launch request and never switch back, so that the original working directory can not be accidentally used by any of the other code.

The tests that intentionally change build directory would be responsible for deferring an os.Chdir to undo the change.

Reply:
I reverted back to the original version of this pr - that passes the build directory down to the build package. We thought about it a lot and we think it's better to limit the scope of buildDir to be only the build stage and leave the delve's working directory untouched.

Made `output` to be buildDir-relative. I have two conflicting thought about this - if the user tries to build a target outside the workspace (e.g. a binary that depends on the libraries in the workspace), writing a file outside the workspace is a bit strange. On the other hand, by defaulting to the package directory (most likely the buildDir), when users start multiple debug sessions concurrently, we will have less chance of multiple delves attempting to work with the same  `__debug_bin` path.

Reply:
Maybe we should finally fix the default path of __debug_bin to be a randomly generated one with ioutil.TempFile.

Reply:
Took a stab at it. Turned out a bit complicated than I thought and I needed to fix the bug [that infers cwd from the `program`](https://github.com/go-delve/delve/blob/09798471003d55918581b8b835c0046e0c2f560a/service/dap/types.go#L108). 
If you think this is better for a separate PR, I am happy to split it.

This change will break user workflow if the test/program to debug has to run in the package directory but users didn't set 'cwd' nor run dlv dap from the package directory. VS Code Go plugin currently runs `dlv dap` from the debugged test or program's directory, so I think it will be ok.

For `test` mode, we may consider to make dlv figure out the package directory (go list?) and use the directory as the default working directory, like `go test`. That will allow `dlv test`  and DAP launch to accept a package path outside the main module (e.g. dependency's test!), but I guess that's quite a niche use case.

Reply:
> For test mode, we may consider to make dlv figure out the package directory (go list?) and use the directory as the default working directory, like go test

`dlv test` already does this, see testCmd in cmd/dlv/cmds/commands.go.

So, now that the output file is in the temp directory, why should buildDir be applied only to the build command?

Reply:
That's great. (Sorry I missed it).
Made a change to infer the package directory as the default cwd for test mode.

Can we use buildDir as the cwd for test mode?

For test mode, gopls simulates the `go.work` mechanics in older versions of go by creating [a workspace module in a temp directory](https://github.com/golang/tools/blob/master/gopls/doc/workspace.md#workspace-module-experimental) and build all targets from the workspace module. The working directory of the test, in this case, is the package directory seen by the go command run from the temporary workspace module directory (buildDir).

For non-test mode, we cannot assume buildDir as the default cwd.

Reply:
> Can we use buildDir as the cwd for test mode?

This is an argument for doing `os.Chidr(buildDir)` as soon as possible and then forgetting about it.

> For non-test mode, we cannot assume buildDir as the default cwd.

Why?

Reply:
> We don't want to do that. Cwd and all related args are can be exposed as 'workspace root'-relative.

The original motivation for this PR was:

> vimspector always launches the debug adapter process from the workspace root.
> Currently building targets outside of the workspace, or outside of the main module isn't
> possible.

This sounds to me like vimspector lacks a feature (letting user specify the working directory for the debugger) and we are writing a workaround for it. So it should minimally do that.
Right now we have that "the working directory = where the go command can be run", this introduces an exception (as well as one more thing that relative paths can be relative to, which we already have too many of). Exceptions are bad, how likely is it that we are going to remember of this particular thing the next time we write something that shells to go?

Even for the relative paths alone, how weird is it that now in launch.json you have some paths that are relative to buildDir (program), some paths that are relative to the working directory of delve (cwd, output, traceDirPath and coreFilePath) and some paths that are relative to cwd (args)?

> In the gopls' experimental module workspace mode, we want to build the test target from the module in the temp directory which users wouldn't know the location, and then switch back to either the package directory (for test) or back to the workspace root or user-specified cwd (for non-test).

This sounds like gopls should either use the `-modfile` compile flag or handle the compilation itself. Delve has never been a fully general compiler wrapper.

Reply:
> The original motivation for this PR was:
> 
> > vimspector always launches the debug adapter process from the workspace root.
> Currently building targets outside of the workspace, or outside of the main module isn't
> possible.
> 
> This sounds to me like vimspector lacks a feature (letting user specify the working directory for the debugger) and we are writing a workaround for it. So it should minimally do that.
> Right now we have that "the working directory = where the go command can be run", this introduces an exception (as well as one more thing that relative paths can be relative to, which we already have too many of). Exceptions are bad, how likely is it that we are going to remember of this particular thing the next time we write something that shells to go?
> 
> Even for the relative paths alone, how weird is it that now in launch.json you have some paths that are relative to buildDir (program), some paths that are relative to the working directory of delve (cwd, output, traceDirPath and coreFilePath) and some paths that are relative to cwd (args)?

I understand your concern about tech debt.

From editor's perspective, workspace root is the directory users are primarily working with by default. buildDir is a hack to work around rare cases where we cannot run `go build` from the workspace root - e.g. multi-module workspace or case-insensitive filesystem or symlink madness. Ideally, most users shouldn't need to worry about `buildDir`.

`cwd`, `traceDirPath`, `coreFilePath` are relevant to the mode where the build step is unnecessary. I found it difficult to explain these without using the workspace root. But `workspace root` is irrelevant to delve. The closest is the delve's process working directory assuming the editor starts delve from the working directory. 

If we do os.Chdir(buildDir), I think the documentation presented to users will be

* `cwd` is interpreted as a path relative to delve's working directory.
* `buildDir` specifies the directory the go command runs  in `debug`, `test` modes. It is irrelevant to other modes. When this is used, it also changes the delve's working directory and, as a result, `cwd` that is computed based on delve's working directory will be affected. 

Note that we still cannot avoid dealing with three directories.

If we take this path (os.Chdir) I think `buildDir` is not the right name because it will be doing beyond what it looks.
We should rename it (maybe `delveWD`? I admit this can be confused with delve's --wd. Other suggestions are welcome), and apply it to all modes. Is it ok with you?

> 
> > In the gopls' experimental module workspace mode, we want to build the test target from the module in the temp directory which users wouldn't know the location, and then switch back to either the package directory (for test) or back to the workspace root or user-specified cwd (for non-test).
> 
> This sounds like gopls should either use the -modfile compile flag or handle the compilation itself. Delve has never been a fully general compiler wrapper.

gopls already utilizes -modfile. The question is how to make Delve behave consistent with gopls. An alternative is to pass -modfile flag as buildFlags and make sure delve use this flag appropriately when shelling out the go command. I think this is as messy as buildDir if not harder.

Reply:
> If we take this path (os.Chdir) I think buildDir is not the right name because it will be doing beyond what it looks.

Yes, it would be `delveWorkDir`, `delveCurrentDir` or something similar, and it would be documented as "directory where Delve will run" and it would be always applied and `cwd`, `traceDirPath`, `coreFilePath` and `program` would always be relative to it.

Reply:
If we go with this approach I think it should be`dlvCwd` to match `cwd`.

To confirm dlv internally will switch dlv to that directory, build in it and resolve all relative paths other than `args` against that directory.
This will help clients that cannot control where dlv is started to have dlv "relaunch" itself in a different directory.
This will NOT help clients that want to resolve paths against the workspace while building outside of the workspace. In particular, vscode-go extension will need to translate relative paths for `cwd`, `program`, `output`,`traceDirPath`, `corFilePath` into absolute workspace-based paths before starting dlv in a directory where we want to build things.

Does this sum up the solution?

Reply:
Yes.

Reply:
gopls has nothing to do with launch.json. It runs bunch of `go list` commands underneath.

gopls had been creating its own world when the experimental workspace module feature was enabled. It merges all the go.mod files found in the workspace into one, create a temporary directory, place the merged go.mod (gopls.mod) in the temporary directory, and invoke all go commands from the temporary directory that contains the unified go.mod file.

This is problematic because other go tools including dlv and go test run by vscode-go extension don't know about it. So, the plan was initially to get the gopls workspace module directory location from gopls (gopls has a command) and make other go tools (including `dlv debug/test`, `go test`, `go list`) run from the gopls workspace module directory when building a go command is necessary. For `dlv`, we were thinking to run `dlv debug/test` from the directory with the right `-wd` flag.

(I think gopls uses `-modfile` for something else (e.g. to avoid corrupting the go.mod file in users' workspace) but I am not 100% certain.)

Good new is: eventually this experimental workspace module got evolved to the go.work proposal. Once `go.work` is adopted (go1.18), I think all of this is unnecessary. We just need to make sure all the commands to run from the directory under the `go.work` file exists. For older versions of Go, I don't know how it would work yet - gopls team is talking about emulating go.work feature from the gopls side. But, I don't know how this will be implemented, and how we address the same problem (i.e. dlv and other tools don't know about this go.work emulation). 

Ability to change the build directory feature will be still useful, when working with `dlv dap` server externally launched.

BTW Will update this PR to  use `dlvCWD`.

Reply:
Done.